proto/go_proto_library.bzl

```diff
@@ -52,10 +53,38 @@ def _go_prefix(ctx):
     prefix = prefix + "/"
   return prefix
 
+def _external_dir(file):
+  """Returns the include dir for external files or empty.
+
+  file: a bazel File object
+
+  Returns:
+    a string path or "" if nothing to do
+
+  Examples to look for
+    'external/some_repo'
+    'bazel-$project/external/some_repo'
+    'bazel-out/local-fastbuild/genfiles/external/some_repo'
+    'bazel-out/local-fastbuild/genfiles/'
+"""
+  toks = file.dirname.split("/")
+  if toks[0] == "external":
+    return "/".join(toks[:2])
+  if not toks[0].startswith("bazel-"):
+    return ""
+  for idx in range(1, len(toks)):
+    if toks[idx] == "external":
+      return "/".join(toks[:idx+2])
```

Comment:
After external is the directory named for the repo
e.g. bazel-proj/external/com_google_foo/...
or bazel-out/local-fastbuild/host/external/com_google_foo/
etc.

So when idx is at "external" you always want to grab 1 more directory to construct the import path.
Following bazel-style for proto and C++, users must still do "root-relative" imports.

So 'import "any.proto";' will not work, it has to be:
import "ptypes/any/any.proto";
or
import "google/protobuf/any.proto";